### PR TITLE
Add carbs to healthkit

### PR DESCRIPTION
**Pull request by dyjaks**
Add carbs to healthkit
Should probably refactor carbs/tmp targets to their own files/subscriptions
Entires show up twice in health but health freeapx knows both don't count? All seems to load properly.
Carbs deleted in health don't delete from FreeAPS. Carbs deleted from freeAPS do delete from health.

Most of HealthKitManager.swift is now duplicate code and should be made better but as I don't know swift or ios I'm hitting a limit here of what I can do 

Carbs in the future don't show. This is an easy fix but I feel is more a design decision. OpenAPS is not carb forward looking so should they be shown?

### CODE REVIEW

**Comment by dyjaks** [comment]
I think this is ready enough for review. Health will sync now with FreeAPS and FreeAPS will read carbs from health but it will not read deleted carbs from health, if that makes sense.

**Comment by ivalkou** [comment]
@DobbyWanKenoby PTAL

**Comment by ivalkou** [comment]
I think it's OK, but we need FAX to Health carbs transfer also

**Comment by dyjaks** [comment]
> I think it's OK, but we need FAX to Health carbs transfer also

Added. Wanted to put in in carb storage but that caused a circular dependency.

**Comment by DobbyWanKenoby** [comment]
I find 2 problem: 

1.  Dublicates of items in Health. https://ibb.co/rvX0Pnf
I think the reason is `id` property of `CarbEntity`. Each time id different. 

I keep checking this problem

2. Manual entered carbs data from FAX don't transfer to Health.

File: FreeAPS/Sources/Services/HealthKit/HealthKitManager.swift [hunk 1]
```diff
@@ -287,17 +413,50 @@ final class BaseHealthKitManager: HealthKitManager, Injectable {
...
+                    glucose: nil,
+                    carb: sample.quantity.doubleValue(for: .gram()).decimal
+                )
+            }
+            .map { sample in
+                CarbsEntry(
+                    _id: sample.healthKitId,
+                    createdAt: sample.date,
+                    carbs: sample.carb as? Decimal ?? 0.0,
+                    enteredBy: nil
```
Comment (DobbyWanKenoby) [inline FreeAPS/Sources/Services/HealthKit/HealthKitManager.swift:436]: entered by "healthkit"

File: FreeAPS/Sources/APS/FetchTreatmentsManager.swift [hunk 2]
```diff
@@ -22,18 +23,23 @@ final class BaseFetchTreatmentsManager: FetchTreatmentsManager, Injectable {
...
             }
-            .sink { carbs, targets in
-                let filteredCarbs = carbs.filter { !($0.enteredBy?.contains(CarbsEntry.manual) ?? false) }
+            .sink { syncDate, carbs, targets, carbsFromHealth in
+                let allCarbs = carbs + carbsFromHealth
+                let filteredCarbs = allCarbs.filter { $0.createdAt > syncDate }
                 if filteredCarbs.isNotEmpty {
+                    debug(.nightscout, "Filtered carbs: \(filteredCarbs)")
                     self.carbsStorage.storeCarbs(filteredCarbs)
+                    self.healthKitManager.saveIfNeeded(carbs: filteredCarbs)
```
Comment (DobbyWanKenoby) [inline FreeAPS/Sources/APS/FetchTreatmentsManager.swift:42]: This code save all data (wich contains data from health) to health. After that we have dublicates in AH.
Please add
`let carbsForHealth = allCarbs.filter { !carbsFromHealth.contains($0) }`
and save only filtered data
self.healthKitManager.saveIfNeeded(carbs: carbsForHealth)

File: FreeAPS.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -1201,6 +1204,7 @@
 			children = (
 				38F3783A2613555C009DB701 /* Config.xcconfig */,
 				3818AA42274BBC1100843DB3 /* ConfigOverride.xcconfig */,
+				29AC4F65277F48C100766404 /* ConfigOverride.xcconfig */,
```
[the remark this answers is not in the document]
Comment (dyjaks) [reply FreeAPS.xcodeproj/project.pbxproj:1205]: Done

File: FreeAPS/Sources/APS/FetchTreatmentsManager.swift [hunk 4]
```diff
@@ -22,18 +23,23 @@ final class BaseFetchTreatmentsManager: FetchTreatmentsManager, Injectable {
...
+                return Publishers.CombineLatest4(
+                    Just(self.carbsStorage.syncDate()),
                     self.nightscoutManager.fetchCarbs(),
-                    self.nightscoutManager.fetchTempTargets()
+                    self.nightscoutManager.fetchTempTargets(),
+                    self.healthKitManager.fetch()
                 ).eraseToAnyPublisher()
             }
-            .sink { carbs, targets in
-                let filteredCarbs = carbs.filter { !($0.enteredBy?.contains(CarbsEntry.manual) ?? false) }
```
[the remark this answers is not in the document]
Comment (dyjaks) [reply FreeAPS/Sources/APS/FetchTreatmentsManager.swift:34]: Must have been a mistake, put bakc.

**Comment by DobbyWanKenoby** [comment]
Houston we have a problem

File: FreeAPS/Sources/Services/WatchManager/WatchManager.swift [hunk 5]
```diff
@@ -216,9 +217,11 @@ extension BaseWatchManager: WCSessionDelegate {
...
 
         if let carbs = message["carbs"] as? Double, carbs > 0 {
-            carbsStorage.storeCarbs([
-                CarbsEntry(createdAt: Date(), carbs: Decimal(carbs), enteredBy: CarbsEntry.manual)
-            ])
+            let carbArray = [
+                CarbsEntry(id: UUID().uuidString, createdAt: Date(), carbs: Decimal(carbs), enteredBy: CarbsEntry.manual)
+            ]
+            carbsStorage.storeCarbs(carbArray)
+            healthKitManager.saveIfNeeded(carbs: carbArray)
```
Comment (DobbyWanKenoby) [inline FreeAPS/Sources/Services/WatchManager/WatchManager.swift:224]: Please read comment, where I tell about `CarbObserver`. 
After that these edits are unnecessary

**Comment by DobbyWanKenoby** [comment]
After a lot of different tests with adding/removing carbs and different situation (such as bad signal), I think, that we shouldn't add carbs to `HealthKit` in `FreeAPS/Sources/APS/FetchTreatmentsManager.swift` `subscribe` method.
All we need for HealthKit support is use `CarbObserver` protocol in `BaseHealthKitManager` like Ivan did it in `BaseNightscoutManager`. After that all new carbs (from FAX, from NS, and others) will send to `HealthKitManager`.

To `BaseHealthKitManager` add:

```
private func subscribe() {
    broadcaster.register(CarbsObserver.self, observer: self)
}
```

To `BaseHealthKitManager.init(resolver:)` add
```
init(resolver: Resolver) {
    // ...
    subscribe()
}
```

At the end of file add

```
extension BaseHealthKitManager: CarbsObserver {
    func carbsDidUpdate(_ carbs: [CarbsEntry]) {
        saveIfNeeded(carbs: carbs)
    }
}
```

**Comment by DobbyWanKenoby** [comment]
Ok finished review.
So, we edit model (`CarbsEntity`) and now have problem with old carbs entities. 
In several days I will release new `MigrationManager` and resolve this problem

**Comment by DobbyWanKenoby** [comment]
New realization in https://github.com/ivalkou/freeaps/pull/188

**Pull request closed without merging**